nncf/torch/tensor_statistics/collectors.py

```diff
@@ -11,243 +11,209 @@
  limitations under the License.
 """
 
-from abc import ABC, abstractmethod
 from collections import deque
-from typing import List, Dict, Set, Tuple, Deque
+from typing import List, Dict, Set, Deque, Union
 
 import torch
 import numpy as np
 
+from nncf.common.tensor_statistics.collectors import OnlineTensorStatisticCollector
+from nncf.common.tensor_statistics.collectors import OfflineTensorStatisticCollector
+from nncf.common.tensor_statistics.collectors import ReductionShape
 from nncf.torch.dynamic_graph.context import no_nncf_trace
 from nncf.torch.tensor_statistics.reduction import min_reduce_like, max_reduce_like, \
-    get_per_channel_history, expand_like, percentile_reduce_like
-from nncf.torch.tensor_statistics.statistics import TensorStatistic, MinMaxTensorStatistic, \
-    MedianMADTensorStatistic, PercentileTensorStatistic
-
-ReductionShape = Tuple[int]
-
-
-class TensorStatisticCollectorBase(ABC):
-    def __init__(self, reduction_shapes: Set[ReductionShape] = None, num_samples: int = None):
-        self._reduction_shapes = reduction_shapes
-        self._enabled = True
-        self._collected_samples = 0
-        self._num_samples = num_samples
-
-    def register_input(self, x: torch.Tensor) -> torch.Tensor:
-        if not self._enabled or \
-                self._num_samples is not None and self._collected_samples >= self._num_samples:
-            return x
-        if self._reduction_shapes is None:
-            self._reduction_shapes = {x.shape}
-        self._register_input(x)
-        self._collected_samples += 1
-        return x
-
-    @abstractmethod
+    get_per_channel_history, expand_like, percentile_reduce_like, get_reduction_shapes
+from nncf.torch.tensor_statistics.statistics import PTMinMaxTensorStatistic
+from nncf.torch.tensor_statistics.statistics import PTMedianMADTensorStatistic
+from nncf.torch.tensor_statistics.statistics import PTPercentileTensorStatistic
+
+
+class PTOfflineTensorStatisticCollector(OfflineTensorStatisticCollector):
     def _register_input(self, x: torch.Tensor):
-        pass
+        with no_nncf_trace():
+            self._samples.append(x.detach().cpu().numpy())
```

Comment:
What happens if I want to calculate several offline statistics for one tensor in the model? Will I get double memory consumption? cc' @vshampor

Reply:
That is right. Currently, the memory consumption will be doubled.